src/test/scala/org/constellation/cluster/E2ETest.scala

```diff
@@ -128,6 +180,51 @@ class E2ETest extends AsyncFlatSpecLike with Matchers with BeforeAndAfterAll wit
 
     val storedSnapshots = allAPIs.map{_.simpleDownload()}
 
+    var numInvalid = 0
+
+    val messagesInChannelWithBlocks = storedSnapshots.head.flatMap{ s =>
+      s.checkpointCache.map{ cache =>
+        val block = cache.checkpointBlock.get
+        val relevantMessages = block.checkpoint.edge.resolvedObservationEdge.data.get.messages
+          .filter{expectedMessages.contains}
+          //.filter{_.signedMessageData.data.channelId == channelId}.filterNot{_ == genesisChannel}
+        val messageParent = relevantMessages.map{_.signedMessageData.data.previousMessageDataHash}.headOption
+        val messageHash = relevantMessages.map{_.signedMessageData.hash}.headOption
+
+        val valid = relevantMessages.map{m =>
+          val isValid = SensorData.validate(
+            m.signedMessageData.data.message
+          ).isSuccess
+          if (!isValid) numInvalid += 1
+          isValid
+        }.headOption
+        (block.soeHash, block.parentSOEHashes, valid, messageParent, messageHash)
+      }
+    }
+
+    // TODO: Duplicate messages appearing sometimes but not others?
+    println(s"Num invalid $numInvalid")
+
+    val ids = messagesInChannelWithBlocks.map{_._1}.zipWithIndex.toMap
+    val msgToBlock = messagesInChannelWithBlocks.flatMap{z => z._5.map{_ -> z._1}}.toMap
```

Comment:
I know this is just test code -- but might be worth making a case class for these. 5 members in a tuple starts is a but unwieldy.

Reply:
It needs to be moved to a separate rendering function as well. This doesn't really belong in the test but it's convenient. Can update.